sdk/lib/cmlib/hiveinit.c

```diff
@@ -362,6 +363,150 @@ HvpInitializeFlatHive(
     return STATUS_SUCCESS;
 }
 
+/**
+ * @name HvpRecoverHeaderFromLog
+ *
+ * Function to recover hive header from log.
+ */
+BOOLEAN CMAPI
+HvpRecoverHeaderFromLog(
+    IN PHHIVE Hive,
+    IN OUT PHBASE_BLOCK *BaseBlock)
+{
+    PHBASE_BLOCK LogBaseBlock;
+    ULONG Offset = 0;
+    BOOLEAN IsSuccess;
+
+    ASSERT(sizeof(HBASE_BLOCK) >= (HSECTOR_SIZE * Hive->Cluster));
+
+    LogBaseBlock = HvpAllocBaseBlockAligned(Hive, TRUE, TAG_CM);
+    if (!LogBaseBlock)
+    {
+        DPRINT1("Allocate base block failed\n");
+        return FALSE;
+    }
+
+    /* Read log file header */
+    IsSuccess = Hive->FileRead(Hive,
+                               HFILE_TYPE_LOG,
+                               &Offset,
+                               LogBaseBlock,
+                               Hive->Cluster * HSECTOR_SIZE);
+
+    if (!IsSuccess)
+    {
+        DPRINT1("Read LOG file failed\n");
+        return FALSE;
+    }
+
+    /* Validate log header */
+    if (!HvpVerifyHiveHeader(LogBaseBlock, HFILE_TYPE_LOG))
+    {
+        DPRINT1("LOG header corrupted\n");
+        // TODO: Self heal
+
+        return FALSE;
+    }
+
+    LogBaseBlock->Type = HFILE_TYPE_PRIMARY;
+    LogBaseBlock->CheckSum = HvpHiveHeaderChecksum(LogBaseBlock);
+    
+    /* Write header to hive */
+    Offset = 0;
+    IsSuccess = Hive->FileWrite(Hive,
+                                HFILE_TYPE_PRIMARY,
+                                &Offset,
+                                LogBaseBlock,
+                                Hive->Cluster * HSECTOR_SIZE);
+
+    if (!IsSuccess)
+    {
+        DPRINT1("Write hive header failed\n");
+        return FALSE;
+    }
+
+    *BaseBlock = LogBaseBlock;
+
+    return TRUE;
+}
+
+/**
+ * @name HvpRecoverDataLog
+ *
+ * Function to recover hive data from log.
+ */
+BOOLEAN CMAPI
```

Comment:
Same suggestion here as for the previous Recover function.

Reply:
Not realy. "recover hive data" and "recover hive header".:smiley: